### PR TITLE
Increase diff contrast and improve gutter behavior

### LINKED ISSUE #2181: Staging line hunks has less visual feedback than old GitHub Desktop
### Description

This is related to (but distinct from) #1688

In the old non-Electron version of GitHub Desktop, there were four visual states the chunk margin could have.

**Selected**
screen shot 2017-07-03 at 2 26 09 pm

(User moves mouse over the margin)

**Selected with mouse hover**
screen shot 2017-07-03 at 2 26 19 pm

(User clicks the mouse)

**Unselected with mouse hover**
screen shot 2017-07-03 at 2 26 28 pm

(User moves the mouse off the margin)

**Unselected**
screen shot 2017-07-03 at 2 26 35 pm

* * *

GitHub Desktop Beta only has three states, even though the way you stage hunks hasn't changed.

**Selected**
screen shot 2017-07-03 at 2 26 56 pm

(User moves mouse over the margin)

**Mouse hover** (actually selected)
screen shot 2017-07-03 at 2 27 07 pm

(User clicks the mouse)

**Mouse hover** (actually unselected) (Wait, did that click do anything? Nothing changed!)
screen shot 2017-07-03 at 2 27 07 pm

(User moves the mouse off the margin)

**Unselected**
screen shot 2017-07-03 at 2 27 16 pm

There's no visual feedback when you click the mouse. It feels bad.

### Version

**GitHub Desktop version:** 0.6.2
**OS version:** macOS 10.12.5

### LINKED ISSUE #2586: Low contrast for colours in diff/Changes
### Description

The red/green diff colours have a very low-visibility contrast, making them  hard to spot. As a related bug-fix/feature request I would like for the app to honour the [Accessibility settings for increased contrast](https://support.apple.com/kb/PH25218?locale=en_GB) which should make borders more pronounced.

See the attached image below where the Desktop app (on macOS) has the least contrast of the tree compared services.

screen shot 2017-08-30 at 15 58 09

### Version

**GitHub Desktop version:** 0.8.0

**OS version:** macOS 10.12.6 (16G29)

### Steps to Reproduce

1. View the Changes of a file.

**Expected behavior:** Good, readable, contrasty colours.

**Actual behavior:** User squints all the time to see the colours. :)

**Reproduces how often:** Every time

### Logs

n/a
#### Additional Information

n/a

### PR DESCRIPTION
**Pull request by niik**
### Increase contrast of red/green backgrounds

We're now using the same colors as GitHub.com

screen shot 2017-10-10 at 18 25 16

(Before on the left, after on the right)

### Darker backgrounds in gutter

We now use an even darker shade in the gutter to highlight the gutter area

screen shot 2017-10-10 at 18 26 35

### Thin border for gutter in read-only mode

Editable
screen shot 2017-10-10 at 18 27 50

Read-only
screen shot 2017-10-10 at 18 27 32

### One more state for hover depending on selected/unselected

colors

This is showing before and after, illustrating the problem of clicking with no visual feedback

### Fix for hunk highlight when hovering on context line

moargutter

# Fix flickering when hovering close to the gutter

flicker

Additionally this moves our `DiffLineGutter` component from directly manipulating the DOM and instead manipulating state and re-rendering.

cc @donokuda for a look but I welcome a code review here as well.

Fixes #2586 
Fixes #2181

### CODE REVIEW

**Comment by j-f1** [comment]
Looks awesome @niik. Note that you have to use “Fixes” before each issue name to auto-close them (`Fixes #2586, fixes #2181`).

File: app/src/ui/diff/diff-line-gutter.tsx [hunk 1]
```diff
@@ -100,34 +127,26 @@ export class DiffLineGutter extends React.Component<IDiffGutterProps, {}> {
    * Indicate whether the current gutter element is selected
    */
   public isIncluded(): boolean {
```
Comment (j-f1) [inline app/src/ui/diff/diff-line-gutter.tsx]: Should this be `isSelected`?

File: app/src/ui/diff/diff-line-gutter.tsx [hunk 2]
```diff
@@ -149,8 +168,17 @@ export class DiffLineGutter extends React.Component<IDiffGutterProps, {}> {
...
     const selectedClass = this.isIncluded() ? selectedLineClass : null
-
-    return classNames('diff-line-gutter', lineClass, selectedClass)
+    const hoverClass = this.state.hover ? hoverCssClass : null
+
+    return classNames(
+      'diff-line-gutter',
+      lineClass,
+      selectedClass,
+      hoverClass,
```
Comment (j-f1) [inline app/src/ui/diff/diff-line-gutter.tsx]: You can remove the `hoverClass` variable and merge this into the object on the next line:

```js
{
  'read-only': this.props.readOnly,
  [hoverCssClass]: this.state.hover,
}
```

File: app/src/ui/diff/diff-line-gutter.tsx [hunk 3]
```diff
@@ -240,19 +256,17 @@ export class DiffLineGutter extends React.Component<IDiffGutterProps, {}> {
...
-        elem.addEventListener('mousedown', this.mouseDownHandler)
-      }
+      elem.addEventListener('mousedown', this.mouseDownHandler)
     } else {
       // this callback fires a second time when the DOM element
-      // is unmounted, so we can use this as a chance to cleanup
-
+      // is unmounted, so we can use this as a chance to cleanup.
+      // We unsubscribe without checking for isIncludeableLine since
+      // that might have changed underneath us
```
Comment (j-f1) [inline app/src/ui/diff/diff-line-gutter.tsx]: `changed underneath us.`? (add a `.`)

File: app/src/ui/diff/diff-mode.ts [hunk 4]
```diff
@@ -13,6 +13,7 @@ const TokenNames: { [key: string]: string | null } = {
   '+': 'diff-add',
   '-': 'diff-delete',
   '@': 'diff-hunk',
+  ' ': 'diff-context',
```
Comment (j-f1) [inline app/src/ui/diff/diff-mode.ts]: Should the token names be an `enum` with string values, so the type is `{ [key: string]: DiffMode | null }`?

File: app/src/ui/diff/diff-line-gutter.tsx (hunk 1 above)
Comment (niik) [reply app/src/ui/diff/diff-line-gutter.tsx]: Yeah, it probably should be. Our terminology here is all over the place but I'm not keen to address that in this PR.

File: app/src/ui/diff/diff-line-gutter.tsx (hunk 2 above)
Comment (niik) [reply app/src/ui/diff/diff-line-gutter.tsx]: That's a neat trick! I'm doing some follow up stuff for this and I'll try to get into the habit of doing this

File: app/src/ui/diff/diff-mode.ts (hunk 4 above)
Comment (niik) [reply app/src/ui/diff/diff-mode.ts]: Yeah, most likely. Our DiffParser does something similar. I'm gonna say that it's out of scope for this PR though.

File: app/src/ui/diff/diff-line-gutter.tsx (hunk 1 above)
Comment (j-f1) [reply app/src/ui/diff/diff-line-gutter.tsx]: I brought it up because all of the uses of the method assign its result to something including `selected`.

**Comment by vojtech-dobes** [comment]
😻

**Pull request merged**